### PR TITLE
Update @woocommerce/components

### LINKED ISSUE #1243: Radio inputs shouldn't be unselectable and there should be an initial value selected
As @senadir posted in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1237#issuecomment-557084336:

> Radio boxes shouldn’t be unselectable and also, there must always be an initial value selected, at least this is what I believe the case should be.
> 
> https://www.nngroup.com/articles/radio-buttons-default-selection/

But currently, all our blocks with radio inputs allow unselecting them. See _Featured Category_, for example:

Peek 2019-11-22 09-39

cc @joshuatf wondering if you have an opinion on that, should we proceed changing the behavior of radio inputs or should we leave them as they are now?

### LINKED ISSUE #1684: Product/Category Controls: wrong aria-expanded attribute used for role="menuitemradio"
imatge

**To Reproduce**
1. Create a post with a _Featured Product_ block.
2. Expand a product with variations.
3. Run an Axe audit and notice there is an error: `Elements must only use allowed ARIA attributes`.

That's because the expanded item uses `role="menuitemradio"` and it has an `aria-expanded` attribute, which is not a valid combination:

https://www.w3.org/TR/wai-aria-1.1/#aria-expanded

https://www.w3.org/TR/wai-aria-1.1/#menuitemradio

### LINKED ISSUE #3603: Tasks to do when updating `@woocommerce/components`
Recently, I pushed a couple of PRs that change the styling of `SearchListControl` and `SearchListItem` (https://github.com/woocommerce/woocommerce-admin/pull/5809 and https://github.com/woocommerce/woocommerce-admin/pull/5901). Some of it was in preparation for the widgets editor (#3228) and to fix some styling issues that we got in recent versions of WP. This issue should be used as a list of the tasks that we will need to do when updating `@woocommerce/components`:

* The operator select that is usually rendered after the list, will no longer need to have a margin-top, since that margin will be included in the list that we get from upstream (previously, it was removed in [this line](https://github.com/woocommerce/woocommerce-admin/pull/5901/files#diff-6674882d8a24592c41ba22ae733ede51a8756959648b7b7266ac3ee62e7cc318L207)). In other words, it should be safe to remove these lines (notice those selectors weren't working any way in recent versions of WP):
  https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/0cffa3e2eca9fa757597a504085e352acb9e0ecf/assets/js/editor-components/product-attribute-term-control/style.scss#L14
  https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/0cffa3e2eca9fa757597a504085e352acb9e0ecf/assets/js/editor-components/product-category-control/style.scss#L14
  https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/0cffa3e2eca9fa757597a504085e352acb9e0ecf/assets/js/editor-components/product-tag-control/style.scss#L14
* We will need to add a new prop (`isCompact={ true }`) to all `SearchListControl` components that are rendered in the sidebar.
* It should be safe to remove this border since that was moved upstream (https://github.com/woocommerce/woocommerce-admin/pull/5901#pullrequestreview-554794141):
  https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/0cffa3e2eca9fa757597a504085e352acb9e0ecf/assets/css/editor.scss#L51-L54
* Since we can no longer rely on the `.components-panel` class, we will need to rewrite the selector for `.components-select-control__input` from this three files
  https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/0cffa3e2eca9fa757597a504085e352acb9e0ecf/assets/js/editor-components/product-attribute-term-control/style.scss#L16
  https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/0cffa3e2eca9fa757597a504085e352acb9e0ecf/assets/js/editor-components/product-category-control/style.scss#L16
  https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/0cffa3e2eca9fa757597a504085e352acb9e0ecf/assets/js/editor-components/product-tag-control/style.scss#L16
* In several files we are creating a custom `SearchListItem` component that is expandable. It will need the `has-count` class name, which was introduced in https://github.com/woocommerce/woocommerce-admin/pull/5809. Ideally, we might also want to explore upstreaming those components as `ExpandableSearchListItem` to avoid duplication.

### LINKED ISSUE #4192: Block widgets: styling issues in blocks using the SearchListControl
_This issue surfaced as part of our audit of the current state of using WooCommerce Blocks as Block Widgets in the new Widgets Editor and Customizer (pca54o-1pT-p2)._

**Note:** This is just a tracking issue as a reminder for us to double-check if the issue has been resolved in the context of Block Widgets once https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4100 is merged. 

In our initial audit, we discovered that there were styling issues with blocks using SearchListControl when they were inside a .components-panel. We created https://github.com/woocommerce/woocommerce-admin/pull/5901 as the first step to resolving the issue and once this in progress PR https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4100 is merged, the issue should be fixed.

Screenshot 2021-05-11 at 23 21 33

### PR DESCRIPTION
**Pull request by Aljullu**
Fixes #3603.
Fixes #4192.
Fixes #1243.
Fixes #1684.

This PR:
* Updates `@woocommerce/components` to version 6.2.0. That required making some changes here and there. They are explained in #3603.
* Creates a `ExpandableSearchListItem` component. There were several controls having a similar logic to make search list items expandable and it could be extracted to its own component. In the future, this could be uplifted to `@woocommerce/components`.
* I refactored and cleaned up some parts of the modified components.

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader

### Screenshots

| Before | After |
| --- | --- |
| imatge | imatge |

### How to test the changes in this Pull Request:

The controls modified in this PR are used in many blocks, so the steps below should be repeated for all of them.

1. Add the block to a page and ensure you can select items in the controls.
2. If there are expandable items (Featured Products or Products by Attribute), make sure you can select inner items.
3. If a count is displayed, make sure it's displayed correctly.
4. Verify that you can do this actions also in the control which are in the sidebar:
imatge
5. Test adding several blocks in the same page, verify controls work independently of each other.

Affected blocks:
* Featured Product Block
* Featured Category Block
* Hand-Picked products Block
* Products by Category Block
* Products by Tag Block
* Products by Attribute Block
* Reviews by Product
* Reviews by Category
* Filter Products by Attribute

### Changelog

> Improved accessibility and styling of the controls of several of ours blocks.

### CODE REVIEW

File: bin/webpack-configs.js [hunk 1]
```diff
@@ -214,7 +214,7 @@ const getMainConfig = ( options = {} ) => {
 		],
 		resolve: {
 			...resolve,
-			extensions: [ '.js', '.ts', '.tsx' ],
+			extensions: [ '.js', '.jsx', '.ts', '.tsx' ],
```
Comment (Aljullu) [inline bin/webpack-configs.js:217]: I don't really know why this is needed but I couldn't get `npm build` to work without it. :confused:

File: assets/js/blocks/reviews/reviews-by-product/edit.js [hunk 2]
```diff
@@ -49,14 +49,13 @@ const ReviewsByProductEditor = ( {
 				countLabel={ sprintf(
 					/* translators: %d is the review count. */
 					_n(
-						'%d Review',
-						'%d Reviews',
+						'%d review',
+						'%d reviews',
```
Comment (Aljullu) [inline assets/js/blocks/reviews/reviews-by-product/edit.js:53]: We didn't have a consistent casing style for strings shown in the count bubble. In 981fa518f5697eb51e623b16b755363636636b77 I made all of them to use sentence casing.

**Comment by Aljullu** [comment]
@opr just a heads-up that this PR is no longer in progress and is now ready for review. :slightly_smiling_face: 

cc @frontdevde since it should fix #4192, which affects the block editor.

**Review by opr — COMMENTED** [review]
### Featured category block
I notice on the Featured Product block, when I'm done changing the setting inside the block and viewing the preview, I can use this pencil icon to return to the settings:
image

This is not present on the Featured Category block, once you've chosen a category it looks like you're stuck with it unless you delete the block and start again.

### Products by Attribute
I noticed this strange behaviour when choosing a different attribute in the inspector controls, the changes weren't reflected in the block and vice/versa. [Loom video](https://www.loom.com/share/bd35423d47f24994a1cb28388b6433f8)

### Filter Product by Attribute
Adding this block shows:  `This block has encountered an error and cannot be previewed.`
I get this error in console too
<details>
<summary>Console error</summary>

```
edit.js:297 Uncaught TypeError: Cannot read property 'isCompact' of undefined
    at renderAttributeControl (edit.js:297)
    at renderEditMode (edit.js:369)
    at Edit (edit.js:385)
    at renderWithHooks (react-dom.development.js:14938)
    at mountIndeterminateComponent (react-dom.development.js:17617)
    at beginWork (react-dom.development.js:18731)
    at HTMLUnknownElement.callCallback (react-dom.development.js:182)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:231)
    at invokeGuardedCallback (react-dom.development.js:286)
    at beginWork$1 (react-dom.development.js:23338)
renderAttributeControl @ edit.js:297
renderEditMode @ edit.js:369
Edit @ edit.js:385
renderWithHooks @ react-dom.development.js:14938
mountIndeterminateComponent @ react-dom.development.js:17617
beginWork @ react-dom.development.js:18731
callCallback @ react-dom.development.js:182
invokeGuardedCallbackDev @ react-dom.development.js:231
invokeGuardedCallback @ react-dom.development.js:286
beginWork$1 @ react-dom.development.js:23338
performUnitOfWork @ react-dom.development.js:22289
workLoopSync @ react-dom.development.js:22265
performSyncWorkOnRoot @ react-dom.development.js:21891
(anonymous) @ react-dom.development.js:11224
unstable_runWithPriority @ react.development.js:2685
runWithPriority$1 @ react-dom.development.js:11174
flushSyncCallbackQueueImpl @ react-dom.development.js:11219
flushSyncCallbackQueue @ react-dom.development.js:11207
discreteUpdates$1 @ react-dom.development.js:22028
discreteUpdates @ react-dom.development.js:803
dispatchDiscreteEvent @ react-dom.development.js:4291
react_devtools_backend.js:2557 The above error occurred in the <Edit> component:
    in Edit (created by WithSpokenMessages(Edit))
    in WithSpokenMessages(Edit) (created by Edit)
    in Edit (created by WithToolbarControls(Edit))
    in WithToolbarControls(Edit) (created by WithInspectorControl(WithToolbarControls(Edit)))
    in WithInspectorControl(WithToolbarControls(Edit)) (created by WithInspectorControl(WithInspectorControl(WithToolbarControls(Edit))))
    in WithInspectorControl(WithInspectorControl(WithToolbarControls(Edit))) (created by WithToolbarControls(WithInspectorControl(WithInspectorControl(WithToolbarControls(Edit)))))
    in WithToolbarControls(WithInspectorControl(WithInspectorControl(WithToolbarControls(Edit)))) (created by WithInspectorControls(WithToolbarControls(WithInspectorControl(WithInspectorControl(WithToolbarControls(Edit))))))
    in WithInspectorControls(WithToolbarControls(WithInspectorControl(WithInspectorControl(WithToolbarControls(Edit)))))
    in Unknown (created by WithDispatch(Component))
    in WithDispatch(Component)
    in Unknown (created by WithMultipleValidation(WithInspectorControls(WithToolbarControls(WithInspectorControl(WithInspectorControl(WithToolbarControls(Edit)))))))
    in WithMultipleValidation(WithInspectorControls(WithToolbarControls(WithInspectorControl(WithInspectorControl(WithToolbarControls(Edit)))))) (created by WithFilters(Edit))
    in WithFilters(Edit) (created by BlockEdit)
    in BlockEdit (created by BlockListBlock)
    in div (created by Block)
    in Block (created by BlockListBlock)
    in BlockCrashBoundary (created by BlockListBlock)
    in BlockListBlock (created by (BlockListBlock))
    in (BlockListBlock) (created by ((BlockListBlock)))
    in ((BlockListBlock)) (created by WithFontSizeInlineStyles(((BlockListBlock))))
    in WithFontSizeInlineStyles(((BlockListBlock))) (created by (WithFontSizeInlineStyles(((BlockListBlock)))))
    in (WithFontSizeInlineStyles(((BlockListBlock)))) (created by WithDefaultAttributes((WithFontSizeInlineStyles(((BlockListBlock))))))
    in WithDefaultAttributes((WithFontSizeInlineStyles(((BlockListBlock))))) (created by WithFilters(BlockListBlock))
    in WithFilters(BlockListBlock) (created by IfCondition(WithFilters(BlockListBlock)))
    in IfCondition(WithFilters(BlockListBlock)) (created by WithDispatch(IfCondition(WithFilters(BlockListBlock))))
    in WithDispatch(IfCondition(WithFilters(BlockListBlock)))
    in Unknown (created by Pure(WithSelect(WithDispatch(IfCondition(WithFilters(BlockListBlock))))))
    in Pure(WithSelect(WithDispatch(IfCondition(WithFilters(BlockListBlock))))) (created by Items)
    in Items (created by BlockListItems)
    in BlockListItems (created by BlockList)
    in div (created by BlockList)
    in BlockList (created by VisualEditor)
    in div (created by WritingFlow)
    in WritingFlow (created by VisualEditor)
    in div (created by VisualEditor)
    in div (created by VisualEditor)
    in VisualEditor (created by Layout)
    in div (created by ForwardRef(InterfaceSkeleton))
    in div (created by ForwardRef(InterfaceSkeleton))
    in div (created by ForwardRef(InterfaceSkeleton))
    in div (created by ForwardRef(InterfaceSkeleton))
    in ForwardRef(InterfaceSkeleton) (created by Layout)
    in Layout (created by Editor)
    in ErrorBoundary (created by Editor)
    in BlockEditorProvider
    in Unknown (created by Context.Consumer)
    in WithRegistryProvider(BlockEditorProvider) (created by EditorProvider)
    in BlockContextProvider (created by EditorProvider)
    in EntityProvider (created by EditorProvider)
    in EntityProvider (created by EditorProvider)
    in EditorProvider
    in Unknown (created by Context.Consumer)
    in WithRegistryProvider(EditorProvider) (created by Editor)
    in DropZoneContextProvider (created by Editor)
    in SlotFillProvider (created by Provider)
    in SlotFillProvider (created by Provider)
    in Provider (created by Editor)
    in StrictMode (created by Editor)
    in Editor

React will try to recreate this component tree from scratch using the error boundary you provided, BlockCrashBoundary.
overrideMethod @ react_devtools_backend.js:2557
logCapturedError @ react-dom.development.js:19662
logError @ react-dom.development.js:19699
callback @ react-dom.development.js:20879
callCallback @ react-dom.development.js:12625
commitUpdateQueue @ react-dom.development.js:12646
commitLifeCycles @ react-dom.development.js:19993
commitLayoutEffects @ react-dom.development.js:22938
callCallback @ react-dom.development.js:182
invokeGuardedCallbackDev @ react-dom.development.js:231
invokeGuardedCallback @ react-dom.development.js:286
commitRootImpl @ react-dom.development.js:22676
unstable_runWithPriority @ react.development.js:2685
runWithPriority$1 @ react-dom.development.js:11174
commitRoot @ react-dom.development.js:22516
finishSyncRender @ react-dom.development.js:21942
performSyncWorkOnRoot @ react-dom.development.js:21928
(anonymous) @ react-dom.development.js:11224
unstable_runWithPriority @ react.development.js:2685
runWithPriority$1 @ react-dom.development.js:11174
flushSyncCallbackQueueImpl @ react-dom.development.js:11219
flushSyncCallbackQueue @ react-dom.development.js:11207
discreteUpdates$1 @ react-dom.development.js:22028
discreteUpdates @ react-dom.development.js:803
dispatchDiscreteEvent @ react-dom.development.js:4291
```
</details>


### General
Some input IDs are duplicated, so using the labels to activate inputs can cause incorrect inputs to be activated. [Loom video](https://www.loom.com/share/64639b5d5f3545cfb30e95a1dcf712ed)

**Comment by Aljullu** [comment]
Thanks for the review @opr!

> Featured category block

We have an open issue for this: #4177. :slightly_smiling_face: 

> Products by Attribute

Good catch! This seems to be an issue in `trunk` too, how does it sound if I open an issue to investigate it? This way we don't increase the scope of this PR.

> Filter Product by Attribute

Fixed in: 49a4b44fd53ef913963c106fcb9a96d711ed3104.

> General

Fixed in: 9da782646246c48067417b3954cf3e80378d73b9.

**Comment by opr** [comment]
> how does it sound if I open an issue to investigate it?

Sounds OK to me 👍🏼

**Review by opr — COMMENTED** [review]
thanks for fixing the other things! They work well now 😎 However I noticed this error when creating a Products by Category block `This block has encountered an error and cannot be previewed.` - to reproduce this error you need to add the block, choose some categories, then try to switch back to editing the block again. I don't think this was introduced with your latest changes, I just missed it in my first round of testing.

https://www.loom.com/share/067134b64d244c7581af59e30d83843e

**Comment by Aljullu** [comment]
@opr that's odd, I'm unable to reproduce. Looking at the video you shared, `SearchListControl` has the old styles. Can you try again after running `npm i`? If you can check if there is any error in the browser console that would be helpful as well, I fixed something similar in this commit 40667170a85fe8a53b9f984d128d8da85710a006, so it might be related.

**Review by opr — APPROVED** [review]
> @opr that's odd, I'm unable to reproduce. Looking at the video you shared, `SearchListControl` has the old styles. Can you try again after running `npm i`? If you can check if there is any error in the browser console that would be helpful as well, I fixed something similar in this commit [4066717](https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/40667170a85fe8a53b9f984d128d8da85710a006), so it might be related.

Weird, I made sure to do `npm i` when I checked this out, but now it's working ok. I restarted my PC between now and then. In that case I'll give this the ✅ - the code changes look good and the testing has been successful. 👏🏼

**Pull request merged**